serac spack pkg: add +asan variant

goxberry opened the pull request:
This commit adds a `+asan` variant to the Serac Spack package to
expose more readily the `ENABLE_ASAN` option to users building Serac
through Spack and Uberenv. This change is probably more convenient for
downstream users of Serac that build it as a dependency of their
projects via Spack/Uberenv.

Resolves #531.

goxberry committed:
serac spack pkg: add +asan variant

This commit adds a `+asan` variant to the Serac Spack package to
expose more readily the `ENABLE_ASAN` option to users building Serac
through Spack and Uberenv. This change is probably more convenient for
downstream users of Serac that build it as a dependency of their
projects via Spack/Uberenv.

Resolves #531.

white238 commented on scripts/spack/packages/serac/package.py:
```diff
@@ -47,6 +47,8 @@ class Serac(CachedCMakePackage, CudaPackage):
             description='Enable runtime safety and debug checks')
     variant('shared',   default=False,
             description='Enable build of shared libraries')
+    variant('asan', default=False,
```
Can you add a conflict if the compiler is not clang or gcc?

goxberry replied:
Good catch! Totally forgot about that restriction.

goxberry replied:
Added checks in [`8916e60f`](https://github.com/LLNL/serac/pull/571/commits/8916e60f5852ee7247e62886bb762fcbebc2d9bc). The checks might be overkill -- comments explain the approach.

goxberry committed:
serac spack pkg: add asan compiler validity checks

This commit adds:

* a blacklist of compilers suspected not to support Address Sanitizer
  to enable some compiler conflict detection at spec concretization
  time

* a whitelist of compilers known to support Address Sanitizer to guard
  against users providing specs using nonstandard compilers and the
  `+asan` variant; this check occurs during the `cmake_args` stage of
  the package and is deferred until Spack executes the package

jamiebramwell committed:
Merge branch 'develop' into feature/oxberry1/serac-asan-variant

jamiebramwell commented:
I'm going to go investigate why this is failing CI

jamiebramwell reviewed (approved):
Thanks @goxberry ! Sorry for the delay. CI woes. 😭
